### PR TITLE
Fixed lifecycle state machine

### PR DESCRIPTION
**Pull request by reseul**
Current code had some issues like moving the contexts just into "background" when suspending (on >=RS1 Windows), etc.

Changed the code a little bit for consistency.
- contexts always go through OnResume, OnLeavingBackground, OnEnteredBackground, OnSuspend now, regardless of the OS version
- Renamed the internal Resumed state to Foreground since it better reflects the "running in foreground" state.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
   "scripts": {
     "test": "jasmine",
     "lint": "eslint .",
-    "flow-check": "flow check"
+    "flow-check": "flow check",
+    "start": "node --max-old-space-size=4096 ./run-packager.js"    
   },
```
Comment (rozele) [inline package.json]: Curious if this commit is intentional. I agree we should probably have some logical `npm start`, but can it go in a separate PR for discussion?

**Comment by rozele** [comment]
@reseul - do the lifecycle events reflect the UWP states here: [https://docs.microsoft.com/en-us/windows/uwp/launch-resume/app-lifecycle](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/app-lifecycle)?

File: ReactWindows/ReactNative.Shared/Common/LifecycleState.cs [hunk 2]
```diff
@@ -17,9 +17,9 @@ public enum LifecycleState
         BeforeResume,
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/Common/LifecycleState.cs]: >BeforeResume [](start = 8, length = 12)

If we're making breaking changes to this public API anyway (I'm comfortable with this), maybe we should just call this "Suspended" to better align with the UWP terminology.

File: ReactWindows/ReactNative.Shared/ReactInstanceManager.cs [hunk 3]
```diff
@@ -257,7 +272,7 @@ public void OnSuspend()
         public void OnEnteredBackground()
         {
             DispatcherHelpers.AssertOnDispatcher();
-            MoveToBackgroundLifecycleState();
+            MoveToBackgroundLifecycleState(false);
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/ReactInstanceManager.cs]: Do you think it would help if we moved the whole state machine to a new class file? ReactInstanceManager is a monolith as is.

**Comment by rozele** [comment]
This is really great. I think there's an opportunity here to get some code out of ReactInstanceManager (especially since we're starting to diverge a bit from ReactAndroid w.r.t. lifecycle management). Thanks for the contribution!

**Comment by reseul** [comment]
Yes, this is what I wanted. Also I wanted those states to work when OS is older.
That's also the reason I renamed the "Resumed" state since it was confusing. You get into a "resumed" state just temporarily (which is actually a background state), until you leave background.
Contexts continue to get OnResume/OLeaveBackground...OnEnterBackground/OnSuspend, similar to that cycle in the doc you sent.



---
In reply to: [363048119](https://github.com/Microsoft/react-native-windows/pull/1616#issuecomment-363048119) [](ancestors = 363048119)

**Comment by reseul** [comment]
Ok, I can move it out, but how exactly would you want that to be? Like a helper class used by ReactInstanceManager? Has to have access to that _currentReactContext, etc.

---
In reply to: [363049229](https://github.com/Microsoft/react-native-windows/pull/1616#issuecomment-363049229) [](ancestors = 363049229)

File: package.json (hunk 1 above)
Comment (reseul) [reply package.json]: Duh, was a mistake. I use this + 2 helper files to start the packager. That's because I've never been able to grok how to do that through existing commands, etc.
I'll remove. 

---
In reply to: [165935883](https://github.com/Microsoft/react-native-windows/pull/1616#discussion_r165935883) [](ancestors = 165935883)

**Comment by reseul** [comment]
@rozele Hi, I pushed the PR feedback fixes.

**Comment by gmangov** [comment]
@reseul @rozele In Skyepe UWP Classic we used an OS implementation of state machine called "Stateless". Came to be quite useful, maybe it can be considered here for implementation of SM.

**Comment by reseul** [comment]
Maybe, it's just that I don't have time right now. I do know I may add regressions when reshuffling stuff, so I tried to keep existing code the same as much as possible.


---
In reply to: [363204818](https://github.com/Microsoft/react-native-windows/pull/1616#issuecomment-363204818) [](ancestors = 363204818)

**Comment by rozele** [comment]
@gmangov - can you file an issue on react-native-windows to investigate Stateless so we don't lose track of this?

---
In reply to: [363206865](https://github.com/Microsoft/react-native-windows/pull/1616#issuecomment-363206865) [](ancestors = 363206865,363204818)

File: ReactWindows/ReactNative.Shared/ReactInstanceManager.cs [hunk 4]
```diff
@@ -582,10 +586,7 @@ private void DetachViewFromInstance(ReactRootView rootView, IReactInstance react
 
             DispatcherHelpers.AssertOnDispatcher();
 
-            if (_lifecycleState == LifecycleState.Resumed)
-            {
-                reactContext.OnSuspend();
-            }
+            _lifecycleStateMachine.SetContext(null);
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/ReactInstanceManager.cs]: Did we lose an `OnSuspend` call here? Sorry @reseul - I know the refactoring makes the diff a bit more complicated.

Comment (reseul) [reply ReactWindows/ReactNative.Shared/ReactInstanceManager.cs]: I moved the logic inside SetContext. State machine brings a new context up to the current state, and a leaving context to OnSuspend.

---
In reply to: [166249288](https://github.com/Microsoft/react-native-windows/pull/1616#discussion_r166249288) [](ancestors = 166249288)

**Comment by reseul** [comment]
@rozele Hey, it seems I "dismissed" your review with a small tweak I added to account for a peculiarity my eyes claim they have seen.
Can you please approve again? also this should be merged, cause there's conflicts I have to fix wit the other PR.

Thanks a lot.

**Comment by reseul** [comment]
@rozele Hi, I played with a new test (locally) as a result of @matthargett request, and I realized we have a bug in case we use the instancemanager.DisposeAsync.
Basically we reset the DispatcherHelpers during OnSuspend, so all "isondispatcher" checks done during DisposeAsync just fail.
I could move the DispatcherHelpers.Reset call to the dispose, are you ok with that? The initialization is idempotent.

File: ReactWindows/ReactNative.Shared/ReactInstanceManager.cs [hunk 5]
```diff
@@ -10,6 +10,9 @@
 using System.Reactive.Disposables;
 using System.Threading;
 using System.Threading.Tasks;
+#if WINDOWS_UWP
+using Windows.Foundation.Metadata;
+#endif
```
Comment (rozele) [inline ReactWindows/ReactNative.Shared/ReactInstanceManager.cs]: I may be missing something, but it doesn't look like this is needed.

**Comment by rozele** [comment]
@reseul -
1. Approved.
2. Can you remove the using statement added to `ReactInstanceManager` if it's not being used?

---
In reply to: [363858167](https://github.com/Microsoft/react-native-windows/pull/1616#issuecomment-363858167) [](ancestors = 363858167)

**Comment by reseul** [comment]
@rozele Every commit invalidates the review :).
